### PR TITLE
[Doc] Lucene inbuilt scalar quantization

### PR DESCRIPTION
**Pull request by naveentatikonda**
### Description
Add documentation for Lucene inbuilt scalar quantization in k-NN plugin.

### Issues Resolved
Closes https://github.com/opensearch-project/documentation-website/issues/6496

### Version
2.16

### Checklist
- [x] By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license and subject to the [Developers Certificate of Origin](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: _search-plugins/knn/knn-vector-quantization.md [hunk 1]
```diff
@@ -17,6 +17,111 @@ OpenSearch supports many varieties of quantization. In general, the level of qua
...
+        }
+      }
+    }
+  }
+}
+```
+{% include copy-curl.html %}
+
+Optionally, you can specify the parameters in `method.parameters.encoder` shown below:
+* `bits` - Determines the size of the quantized vector after quantizing the input float vectors. For instance, 7 bits will quantize input float vectors into 7-bit integer vectors. As of OpenSearch 2.16, only `7` bits are supported. Default value is `7`.
```
Comment (jmazanec15) [inline _search-plugins/knn/knn-vector-quantization.md:64]: Do we need to mention this now if its only 7?

Comment (naveentatikonda) [reply _search-plugins/knn/knn-vector-quantization.md]: removed

**Comment by naveentatikonda** [comment]
@kolchfa-aws Can you pls review this PR. Thanks!

**Comment by kolchfa-aws** [comment]
@naveentatikonda Sure, will do!

File: _search-plugins/knn/knn-vector-quantization.md [hunk 2]
```diff
@@ -17,6 +17,110 @@ OpenSearch supports many varieties of quantization. In general, the level of qua
...
+      }
+    }
+  }
+}
+```
+{% include copy-curl.html %}
+
+Optionally, you can specify the parameters in `method.parameters.encoder` shown below:
+* `confidence_interval` - used to compute the `minQuantile` and `maxQuantile` parameters which are used to quantize the vectors. The accepted values are: 
+  - It can be any value between and including `0.9` to `1.0`. For example, if we set it to `0.9` then it will consider the middle 90% of the vector values for computing the min and max Quantiles excluding the minimum and maximum 5% of the values.
```
Comment (kolchfa-aws) [inline _search-plugins/knn/knn-vector-quantization.md:65]: So, to make sure, valid values are 0.9--1.0, inclusive (for static computation), or 0 (for dynamic computation)?

File: _search-plugins/knn/knn-vector-quantization.md [hunk 3]
```diff
@@ -17,6 +17,110 @@ OpenSearch supports many varieties of quantization. In general, the level of qua
...
+
+There are no changes to ingestion or query mapping and no range limitations for the input vectors. 
+
+### Memory estimation
+
+In the best-case scenario, 7-bit vectors produced by the Lucene scalar quantizer require 25% of the memory that 32-bit vectors require.
+
+#### HNSW memory estimation
+
+The memory required for Hierarchical Navigable Small Worlds (HNSW) is estimated to be `1.1 * (dimension + 8 * M)` bytes/vector.
```
Comment (kolchfa-aws) [inline _search-plugins/knn/knn-vector-quantization.md:116]: What is `M`?

File: _search-plugins/knn/knn-vector-quantization.md (hunk 2 above)
Comment (naveentatikonda) [reply _search-plugins/knn/knn-vector-quantization.md]: yes, that's correct. Also, users can skip this parameter which will be computed as shown under the default case

File: _search-plugins/knn/knn-vector-quantization.md (hunk 3 above)
Comment (naveentatikonda) [reply _search-plugins/knn/knn-vector-quantization.md]: M is max number of connections. This isn't a new parameter.

File: _search-plugins/knn/knn-vector-quantization.md [hunk 4]
```diff
@@ -15,19 +15,18 @@ OpenSearch supports many varieties of quantization. In general, the level of qua
...
-The scalar quantizer in lucene dynamically quantizes the input 32-bit floating-point vectors into 7-bit integer vectors in each segment using the minQuantile and maxQuantile computed using the `confidence_interval` parameter.
+Starting with version 2.16, the k-NN plugin supports built-in scalar quantization for the Lucene engine. Unlike the [Lucene byte vector]({{site.url}}{{site.baseurl}}/field-types/supported-field-types/knn-vector#lucene-byte-vector), which requires you to compress vectors before ingesting the documents, the Lucene scalar quantizer compresses input vectors within OpenSearch during ingestion. The Lucene scalar quantizer converts 32-bit floating-point input vectors into 7-bit integer vectors in each segment using the minimum and maximum quantiles computed based on the `confidence_interval` parameter.
```
Comment (naveentatikonda) [inline _search-plugins/knn/knn-vector-quantization.md:22]: We shouldn't use "compress" here, "quantize" is the correct word

File: _search-plugins/knn/knn-vector-quantization.md [hunk 5]
```diff
@@ -15,7 +15,113 @@ OpenSearch supports many varieties of quantization. In general, the level of qua
...
+Starting with k-NN plugin version 2.9, you can use `byte` vectors with the Lucene engine in order to reduce the amount of required memory. This requires quantizing the vectors outside of OpenSearch before ingesting them into an OpenSearch index. For more information, see [Lucene byte vector]({{site.url}}{{site.baseurl}}/field-types/supported-field-types/knn-vector#lucene-byte-vector).
+
+## Lucene scalar quantization
+
+Starting with version 2.16, the k-NN plugin supports built-in scalar quantization for the Lucene engine. Unlike the [Lucene byte vector]({{site.url}}{{site.baseurl}}/field-types/supported-field-types/knn-vector#lucene-byte-vector), which requires you to quantize vectors before ingesting the documents, the Lucene scalar quantizer quantizes input vectors within OpenSearch during ingestion. The Lucene scalar quantizer converts 32-bit floating-point input vectors into 7-bit integer vectors in each segment using the minimum and maximum quantiles computed based on the `confidence_interval` parameter.
```
Comment (kolchfa-aws) [inline _search-plugins/knn/knn-vector-quantization.md:22]:
```suggestion
Starting with version 2.16, the k-NN plugin supports built-in scalar quantization for the Lucene engine. Unlike the [Lucene byte vector]({{site.url}}{{site.baseurl}}/field-types/supported-field-types/knn-vector#lucene-byte-vector), which requires you to quantize vectors before ingesting the documents, the Lucene scalar quantizer quantizes input vectors within OpenSearch during ingestion. The Lucene scalar quantizer converts 32-bit floating-point input vectors into 7-bit integer vectors in each segment using the minimum and maximum quantiles computed based on the [`confidence_interval`](#confidence-interval) parameter.
```

File: _search-plugins/knn/knn-vector-quantization.md [hunk 6]
```diff
@@ -15,7 +15,113 @@ OpenSearch supports many varieties of quantization. In general, the level of qua
...
+
+There are no changes to ingestion or query mapping and no range limitations for the input vectors. 
+
+### Memory estimation
+
+In the ideal scenario, 7-bit vectors created by the Lucene scalar quantizer use only 25% of the memory required by 32-bit vectors.
+
+#### HNSW memory estimation
+
+The memory required for the Hierarchical Navigable Small World (HNSW) graph can be estimated as `1.1 * (dimension + 8 * M)` bytes/vector, where `M` is the maximum number of bi-directional links created for each element during the construction of the graph.
```
Comment (kolchfa-aws) [inline _search-plugins/knn/knn-vector-quantization.md:118]:
```suggestion
The memory required for the Hierarchical Navigable Small World (HNSW) graph can be estimated as `1.1 * (dimension + 8 * M)` bytes/vector, where `M` is the maximum number of bidirectional links created for each element during the construction of the graph.
```

File: _search-plugins/knn/knn-vector-quantization.md [hunk 7]
```diff
@@ -15,7 +15,113 @@ OpenSearch supports many varieties of quantization. In general, the level of qua
...
+
+## Lucene scalar quantization
+
+Starting with version 2.16, the k-NN plugin supports built-in scalar quantization for the Lucene engine. Unlike the [Lucene byte vector]({{site.url}}{{site.baseurl}}/field-types/supported-field-types/knn-vector#lucene-byte-vector), which requires you to quantize vectors before ingesting the documents, the Lucene scalar quantizer quantizes input vectors within OpenSearch during ingestion. The Lucene scalar quantizer converts 32-bit floating-point input vectors into 7-bit integer vectors in each segment using the minimum and maximum quantiles computed based on the [`confidence_interval`](#confidence-interval) parameter. During search, the query vector is quantized in each segment using the segment’s minimum and maximum quantiles in order to compute the distance between the query vector and the segment’s quantized input vectors.
```
Comment (natebower) [inline _search-plugins/knn/knn-vector-quantization.md:22]:
```suggestion
Starting with version 2.16, the k-NN plugin supports built-in scalar quantization for the Lucene engine. Unlike the [Lucene byte vector]({{site.url}}{{site.baseurl}}/field-types/supported-field-types/knn-vector#lucene-byte-vector), which requires you to quantize vectors before ingesting the documents, the Lucene scalar quantizer quantizes input vectors in OpenSearch during ingestion. The Lucene scalar quantizer converts 32-bit floating-point input vectors into 7-bit integer vectors in each segment using the minimum and maximum quantiles computed based on the [`confidence_interval`](#confidence-interval) parameter. During search, the query vector is quantized in each segment using the segment's minimum and maximum quantiles in order to compute the distance between the query vector and the segment's quantized input vectors. 
```

File: _search-plugins/knn/knn-vector-quantization.md [hunk 8]
```diff
@@ -15,7 +15,113 @@ OpenSearch supports many varieties of quantization. In general, the level of qua
...
+
+Quantization can decrease the memory footprint by a factor of 4 in exchange for some loss in recall. Additionally, quantization slightly increases disk usage because it requires storing both the raw input vectors and the quantized vectors.
+
+### Using Lucene scalar quantization
+
+To use the Lucene scalar quantizer, set the k-NN vector field’s `method.parameters.encoder.name` to `sq` when creating a k-NN index:
```
Comment (natebower) [inline _search-plugins/knn/knn-vector-quantization.md:28]:
```suggestion
To use the Lucene scalar quantizer, set the k-NN vector field's `method.parameters.encoder.name` to `sq` when creating a k-NN index:
```

File: _search-plugins/knn/knn-vector-quantization.md [hunk 9]
```diff
@@ -15,7 +15,113 @@ OpenSearch supports many varieties of quantization. In general, the level of qua
...
+  }
+}
+```
+{% include copy-curl.html %}
+
+### Confidence interval
+
+Optionally, you can specify the `confidence_interval` parameter in the `method.parameters.encoder` object.
+The `confidence_interval` is used to compute the minimum and maximum quantiles in order to quantize the vectors:
+- If you set the `confidence_interval` to a value in `0.9` to `1.0` range, inclusive, then the quantiles are calculated statically. For example, setting the `confidence_interval` to `0.9` specifies to compute the minimum and maximum quantiles based on the middle 90% of the vector values, excluding the minimum 5% and maximum 5% of the values.
```
Comment (natebower) [inline _search-plugins/knn/knn-vector-quantization.md:66]:
```suggestion
- If you set the `confidence_interval` to a value in the `0.9` to `1.0` range, inclusive, then the quantiles are calculated statically. For example, setting the `confidence_interval` to `0.9` specifies to compute the minimum and maximum quantiles based on the middle 90% of the vector values, excluding the minimum 5% and maximum 5% of the values. 
```

Comment (natebower) [inline _search-plugins/knn/knn-vector-quantization.md:66]: "middle" => "median"?, "minimum" => "lowest"?, "maximum" => "highest"?

File: _search-plugins/knn/knn-vector-quantization.md [hunk 10]
```diff
@@ -15,7 +15,113 @@ OpenSearch supports many varieties of quantization. In general, the level of qua
...
+
+Optionally, you can specify the `confidence_interval` parameter in the `method.parameters.encoder` object.
+The `confidence_interval` is used to compute the minimum and maximum quantiles in order to quantize the vectors:
+- If you set the `confidence_interval` to a value in `0.9` to `1.0` range, inclusive, then the quantiles are calculated statically. For example, setting the `confidence_interval` to `0.9` specifies to compute the minimum and maximum quantiles based on the middle 90% of the vector values, excluding the minimum 5% and maximum 5% of the values. 
+- Setting `confidence_interval` to `0` specifies to compute the quantiles dynamically, which involves oversampling and additional computations performed on the input data.
+- When `confidence_interval` is not set, it is computed based on the vector dimension $$d$$ using the formula $$max(0.9, 1 - \frac{1}{1 + d})$$.
+
+Lucene scalar quantization is applied only to `float` vectors. If you change the default value of the `data_type` parameter from `float` to `byte` or any other type when mapping a [k-NN vector]({{site.url}}{{site.baseurl}}/field-types/supported-field-types/knn-vector/), the request is rejected.
```
Comment (natebower) [inline _search-plugins/knn/knn-vector-quantization.md:70]:
```suggestion
Lucene scalar quantization is applied only to `float` vectors. If you change the default value of the `data_type` parameter from `float` to `byte` or any other type when mapping a [k-NN vector]({{site.url}}{{site.baseurl}}/field-types/supported-field-types/knn-vector/), then the request is rejected.
```

File: _search-plugins/knn/knn-vector-quantization.md [hunk 11]
```diff
@@ -15,7 +15,113 @@ OpenSearch supports many varieties of quantization. In general, the level of qua
...
+{: .warning}
+
+The following example method definition specifies the Lucene `sq` encoder with the `confidence_interval` set to `1.0`. This `confidence_interval` specifies to consider all the input vectors for computing the minimum and maximum quantiles. Vectors are quantized to 7 bits by default:
```
Comment (natebower) [inline _search-plugins/knn/knn-vector-quantization.md:73]:
```suggestion
The following example method definition specifies the Lucene `sq` encoder with the `confidence_interval` set to `1.0`. This `confidence_interval` specifies to consider all the input vectors when computing the minimum and maximum quantiles. Vectors are quantized to 7 bits by default:
```

**Review by natebower — COMMENTED** [review]
@kolchfa-aws @naveentatikonda Please see my comments and changes and let me know if you have any questions. Thanks!

**Review by kolchfa-aws — APPROVED** [review]
Thank you, @naveentatikonda!

File: _search-plugins/knn/knn-vector-quantization.md (hunk 9 above)
Comment (kolchfa-aws) [reply _search-plugins/knn/knn-vector-quantization.md]: Definitely "middle". I would keep as is. @naveentatikonda WDYT?

Comment (natebower) [reply _search-plugins/knn/knn-vector-quantization.md]: This is definitely not my area of expertise, so that's fine with me :-)

Comment (naveentatikonda) [reply _search-plugins/knn/knn-vector-quantization.md]: Yes, we should use middle

**Pull request merged**